File: Blind 75/08 - Tries/212-word-search-II.py
'''
Given an m x n board of characters and a list of strings words, return all words on the board.

Each word must be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.
'''
from time import time
from typing import List


class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def add_word(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.is_end = True


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie()
        for w in words:
            trie.add_word(w)

        ROWS, COLS = len(board), len(board[0])
        res = set()

        def dfs(r, c, node, word):
            if (r < 0 or c < 0
                or r >= ROWS or c >= COLS
                    or board[r][c] not in node.children or board[r][c] == '#'):
                return

            # Add letter to word and increment node
            word += board[r][c]
            node = node.children[board[r][c]]
            if node.is_end:
                res.add(word)

            # Store current letter for backtracking
            temp = board[r][c]
            board[r][c] = '#'

            dfs(r - 1, c, node, word)
            dfs(r + 1, c, node, word)
            dfs(r, c - 1, node, word)
            dfs(r, c + 1, node, word)

            # Reset to original letter
            board[r][c] = temp

        for r in range(ROWS):
            for c in range(COLS):
                dfs(r, c, trie.root, '')

        return list(res)

    def reference(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie()
        for w in words:
            trie.add_word(w)

        ROWS, COLS = len(board), len(board[0])
        res, visit = set(), set()

        def dfs_reference(r, c, node, word):
            if (r < 0 or c < 0
                or r == ROWS or c == COLS
                    or (r, c) in visit
                    or board[r][c] not in node.children):
                return

            visit.add((r, c))
            node = node.children[board[r][c]]
            word += board[r][c]
            if node.is_end:
                res.add(word)

            dfs_reference(r - 1, c, node, word)
            dfs_reference(r + 1, c, node, word)
            dfs_reference(r, c - 1, node, word)
            dfs_reference(r, c + 1, node, word)
            visit.remove((r, c))

        for r in range(ROWS):
            for c in range(COLS):
                dfs_reference(r, c, trie.root, '')

        return list(res)

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.findWords(case[0], case[1]))
                else:
                    self.findWords(case[0], case[1])
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case[0], case[1]))
                else:
                    self.reference(case[0], case[1])
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        (
            [['o', 'a', 'a', 'n'], ['e', 't', 'a', 'e'],
             ['i', 'h', 'k', 'r'], ['i', 'f', 'l', 'v']],
            ['oath', 'pea', 'eat', 'rain'],
        ),
        (
            [['a', 'b'], ['c', 'd']],
            ['abcb']
        )
    ]
    test.quantify(test_cases)

File: Blind 75/08 - Tries/JavaScript/212-word-search-II.js
const { performance } = require("perf_hooks");

class TrieNode {
  constructor() {
    this.children = {};
    this.isEnd = false;
  }
}

class TrieNodeRef {
  constructor() {
    this.children = {};
    this.word = "";
  }
}

class Trie {
  constructor() {
    this.root = new TrieNode();
  }

  add(word) {
    let node = this.root;
    for (const c of word) {
      if (!(c in node.children)) node.children[c] = new TrieNode();
      node = node.children[c];
    }

    node.isEnd = true;
  }
}

class TrieRef {
  constructor(words) {
    this.root = new TrieNodeRef();
    words.forEach((word) => this.insert(word));
  }

  /* Time O(N) | Space O(N) */
  insert(word, node = this.root) {
    for (const char of word) {
      const child = node.children[char] || new TrieNodeRef();

      node.children[char] = child;

      node = child;
    }

    node.word = word;
  }

  /* Time O((ROWS * COLS) * (4 * (3 ^ (WORDS - 1)))) | Space O(N) */
  searchBoard(board, node = this.root, words = []) {
    const [rows, cols] = [board.length, board[0].length];

    for (let row = 0; row < rows; row++) {
      for (let col = 0; col < cols; col++) {
        this.dfs(board, row, rows, col, cols, node, words);
      }
    }

    return words;
  }

  dfs(board, row, rows, col, cols, node, words) {
    const char = board[row][col];
    const child = node.children[char] || null;

    if (this.canSkip(char, child)) return;

    node = child;
    this.checkWord(node, words);
    this.backTrack(board, row, rows, col, cols, node, words);
  }

  canSkip(char, child) {
    const hasSeen = char === "#";
    const isMissingChild = !child;

    return hasSeen || isMissingChild;
  }

  checkWord(node, words) {
    if (!node.word.length) return;

    words.push(node.word);
    node.word = "";
  }

  backTrack(board, row, rows, col, cols, node, words) {
    const char = board[row][col];

    board[row][col] = "#";

    for (const [_row, _col] of this.getNeighbors(row, rows, col, cols)) {
      this.dfs(board, _row, rows, _col, cols, node, words);
    }

    board[row][col] = char;
  }

  getNeighbors(row, rows, col, cols) {
    return [
      [row - 1, col],
      [row + 1, col],
      [row, col - 1],
      [row, col + 1],
    ].filter(([_row, _col]) => !this.isOutOfBounds(_row, rows, _col, cols));
  }

  isOutOfBounds(row, rows, col, cols) {
    const isRowOut = row < 0 || rows <= row;
    const isColOut = col < 0 || cols <= col;

    return isRowOut || isColOut;
  }
}

class Solution {
  method(board, words) {
    let trie = new Trie();
    for (const word of words) {
      trie.add(word);
    }

    const [ROWS, COLS] = [board.length, board[0].length];
    let res = new Set();

    const dfs = (r, c, node, word) => {
      if (
        r < 0 ||
        c < 0 ||
        r >= ROWS ||
        c >= COLS ||
        !(board[r][c] in node.children) ||
        board[r][c] === "#"
      ) {
        return;
      }

      word += board[r][c];
      node = node.children[board[r][c]];
      if (node.isEnd) {
        res.add(word);
      }

      const temp = board[r][c];
      board[r][c] = "#";

      dfs(r + 1, c, node, word);
      dfs(r - 1, c, node, word);
      dfs(r, c + 1, node, word);
      dfs(r, c - 1, node, word);

      board[r][c] = temp;
    };

    for (let r = 0; r < ROWS; r++) {
      for (let c = 0; c < COLS; c++) {
        dfs(r, c, trie.root, "", board);
      }
    }

    return [...res];
  }

  reference(board, words) {
    return new TrieRef(words).searchBoard(board);
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [
      ["o", "a", "a", "n"],
      ["e", "t", "a", "e"],
      ["i", "h", "k", "r"],
      ["i", "f", "l", "v"],
    ],
    ["oath", "pea", "eat", "rain"],
  ],
  [
    [
      ["a", "b"],
      ["c", "d"],
    ],
    ["abcb"],
  ],
];
test.quantify(testCases);
